sdk/include/opentelemetry/sdk/logs/exporter.h

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <memory>
+#include <vector>
+#include "opentelemetry/logs/log_record.h"
+#include "opentelemetry/sdk/logs/processor.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace logs
+{
+/**
+ * ExportResult is returned as result of exporting a batch of Log Records.
+ */
+enum class ExportResult
+{
+  // The batch was exported successfully
+  kSuccess = 0,
+  // The batch was exported unsuccessfully and was dropped
+  kFailure
+};
+
+/**
+ * LogExporter defines the interface that log exporters must implement.
+ */
+class LogExporter
+{
+public:
+  virtual ~LogExporter() = default;
+
+  /**
+   * Exporters that implement this should typically format each LogRecord into the format
+   * required by the exporter destination (e.g. JSON), then send the LogRecord to the exporter.
+   * The exporter may retry logs a maximum of 3 times before dropping and returning kFailure.
+   * If this exporter is already shut down, should return kFailure.
+   * @param records: a vector of unique pointers to log records
+   * @returns an ExportResult code (whether export was success or failure)
+   *
+   * TODO: This method should not block indefinitely. Should abort within timeout.
+   */
+  virtual ExportResult Export(
+      const std::vector<std::unique_ptr<opentelemetry::logs::LogRecord>> &records) noexcept = 0;
```

Comment:
Do we really want to pass-by-reference here?

1. Is it possible we want `LogRecord`s to go to multiple exporters?
2. Shouldn't the lifecycle of `LogRecord` destruction be fully owned by the `Exporter` at this point?

Reply:
I think either the processor or exporter could be responsible for destroying the log records, since there is only 1 configured exporter per processor. If the processor were to destroy it,  I think either having a pass by value or pass by reference here would be fine, since neither the processor nor exporter will be making changes to the log record. I'm leaning towards a pass by reference here to prevent unnecessary copying?

Reply:
I was suggesting we use a move here (`auto&&` vs `const auto&`).

Given this is expected to be used from a single thread, not important, was just curious what the lifecycle thought process was.

Reply:
Any reason not to use a `nostd::span` instead of the `std::vector`? That's how the [exporter interface for the tracing SDK](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/sdk/include/opentelemetry/sdk/trace/exporter.h#L53) does it. I think we should aim for a consistent interface among the SDK parts.

I'm not sure about the `std::unique_ptr` though. @jsuereth rightly mentions that there might be several exporters (users can have several processors per logger), I think this will be a popular use case. I see a few possible solutions here:
1. Making copies of log records for each processor/exporter, and passing those as `std::unique_ptr`. This is possibly expensive and not ideal.
2. Keep one instance of each log record and pass it to each processor/exporter as a `std::shared_ptr`. Probably the most pragmatic solution in terms of implementation effort.
3. Following a similar approach to the tracing SDK, where a `Recordable` is used (see here for [the idea](https://github.com/open-telemetry/opentelemetry-cpp/issues/2)). Probably the most performant, but also the most implementation effort.

Reply:
No reason, so I will change `vector` to `span`. Thanks for the catch.

The `unique_ptr` is used here to keep consistency with the fact that the current logging SDK uses `std::move` on the `LogRecord` from the Logger to the processor. The `std::move` was the original idea since the SDK initially only has support for [1 processor](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/sdk/src/logs/logger_provider.cc#L70) so far.

In the next PR, we will be changing the Logger API/SDK to create a `shared_ptr LogRecord` instead of a plain `LogRecord` in the Logger API, which will then be passed to the Logger SDK, which can then be shared with all the processors/exporters. I think this would match suggested option # 2. 

I looked into the Recordable for traces, and based on what I've learned, I'm not sure it could be completed before our last day. Would it be possible to flag this as an issue to tackle in the future?

Reply:
Let's discuss more on `shared_ptr` in the follow on PR (as you're not migrating it to shared_ptr here).  It'd be nice if we can have consistent memory-management performance here, and `shared_ptr` can create inconsistent performance.   There's a LOT of interactions and flow of data that we expect in a logging API and it'd be nice if we have a benchmark to evaluate realistic loads + performance characteristics to better understand the tradeoffs we're making here.   It sounds simple, but the runtime characteristics can change drastically in large systems.

Reply:
Also, for the record @pyohannes #3 is what I think we'll need to implement, and it is pretty sizeable.

I think you should flag it as an issue.  What you have is fine for prototyping and setting up benchmarks and filling out more of the project, but will need to be revisited.

You should open an issue and we should avoid spinning up too many exporters prior to resolving it, though, as that would increase the amount of code we'd need to refactor.

Reply:
Flagged in #412